service/header/sync.go

```diff
@@ -330,3 +427,120 @@ func (s *Syncer) getHeaders(ctx context.Context, start, amount uint64) ([]*Exten
 
 	return out, nil
 }
+
+// heightSub provides a way to wait until a specific height becomes available and synced
+type heightSub struct {
+	// usually we don't attach context to structs, but this is an exception
+	// as GetByHeight and ProvideHeights needs to access it
+	ctx       context.Context
+	cancel    context.CancelFunc
+	store     Store
+	height    uint64
+	provideCh chan []*ExtendedHeader
+	reqsCh    chan *heightReq
+	reqs      map[uint64][]chan *heightResp
+}
+
+type heightReq struct {
+	resp   chan *heightResp
+	height uint64
+}
+
+type heightResp struct {
+	header *ExtendedHeader
+	err    error
+}
+
+func newHeightSub(store Store) *heightSub {
+	return &heightSub{
+		store:     store,
+		provideCh: make(chan []*ExtendedHeader, 4),
+		reqsCh:    make(chan *heightReq, 4),
+		reqs:      make(map[uint64][]chan *heightResp),
+	}
+}
+
+func (hs *heightSub) Start() {
+	hs.ctx, hs.cancel = context.WithCancel(context.Background())
+	go hs.subLoop()
+}
+
+func (hs *heightSub) Stop() {
+	hs.cancel()
+}
+
+func (hs *heightSub) GetByHeight(ctx context.Context, height uint64) (*ExtendedHeader, error) {
+	h, err := hs.store.GetByHeight(ctx, height)
+	if err != ErrNotFound {
+		return h, err
+	}
+
+	resp := make(chan *heightResp, 1)
+	select {
+	case hs.reqsCh <- &heightReq{resp, height}:
+		select {
+		case resp := <-resp:
+			return resp.header, resp.err
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case <-hs.ctx.Done():
+			return nil, hs.ctx.Err()
+		}
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-hs.ctx.Done():
+		return nil, hs.ctx.Err()
+	}
+}
+
+func (hs *heightSub) ProvideHeights(ctx context.Context, headers ...*ExtendedHeader) {
+	select {
+	case hs.provideCh <- headers:
+	case <-ctx.Done():
+	case <-hs.ctx.Done():
+	}
+}
+
+func (hs *heightSub) subLoop() {
```

Comment:
Can you document this?

Reply:
Added a simple comment, not sure what else to say there.

Reply:
LGTM, resolving.